text/projects/Rafal_Kilar/Hunt-Szymanski.tex

```diff
@@ -0,0 +1,156 @@
+
+Załóżmy teraz, że takie minimalne $j$ istnieje. $A[1..i+1]$ i $B[1..j]$ mają wspólny podciąg długości $k$ - wspólne podciąg długości $k-1$ $A[1..i]$ i $B[1..T_{i,k-1}]$ powiększone o $A[i+1]$ i $B[j]$, więc $T_{i+1,k} \le j$. 
+
+Załóżmy niewprost, że $T_{i+1,k} < j$. Z lematu 2. mamy $T_{i,k-1} < T_{i+1,k}$, więc ostatni znak wspólnego podciągu $A[1..i+1]$ i $B[1..T_{i+1,k}]$ długości $k$ nie jest równy $A[i+1]$. $A[1..i]$ i $B[1..T_{i+1,k}]$ muszą również mieć podciąg długości $k$, więc $T_{i,k} \le T_{i+1,k}$. Z lematu 2. mamy więc równość $T_{i, k} = T_{i+1,k}$. Z założenia niewprost i wymogu $T_{i+1,k} < j \le T_{i,k}$ mamy $T_{i,k} \neq T_{i+1,k}$, co daje nam sprzeczność i kończy dowód.
+
+\begin{algorithm}
+\caption{Algorytm HS}\label{alg:hs}
+\begin{algorithmic}
+\For{$i = 1 .. m$}
+    \State $MATCHLIST[i] = \{j_1, j_2, \dots, j_p\} \text{ such that $j_1 > \dots > j_p$ and $A[i] = B[j_k]$}$
+    \State $THRESH[i] = n + 1$
+\EndFor
+\State $THRESH[0] = 0$
+\State $LINK[0] = 0$
+\For{$i = 1 .. m$}
+    \For{$j \text{ in } MATCHLIST[i]$}
+        \State $\text{find $k$ such that} THRESH[k-1] < j < THRESH[k]$
+        \If{$j < THRESH[k]$}
+            \State $THRESH[k] = j$
+            \State $LINK[k] = newnode(i, j, LINK[k-1]$
+        \EndIf
+    \EndFor
+\EndFor
+\State recover LCS in reverse using the $LINK$ array
+\end{algorithmic}
+\end{algorithm}
+
+W naszym algorytmie będziemy wyznaczać kolejne wiersze $T_{i,1}, \dots$.
+
+Algorytm będzie iterował się po kolejnych znakach $A$. Na początku każdej iteracji tablica $THRESH$ będzie zawierała kolejne wartości $T_{i,1}, T_{i,2}, \dots$. Lista $MATCHLIST[i]$ zawiera wszystkie indeksy $j$ takie, że $B[j] = A[i]$ w malejącej kolejności. 
+
+Załóżmy, że dla w iteracji $i$ algorytm rozważa dopasowania $A[i]$ do $B[j_1], \dots, B[j_p]$ pomiędzy wartościami progowymi, to jest $THRESH[k-1] = T_{i-1,k-1} < j_1 < \dots < j_p \le T_{i-1, k} = THRESH[k]$. Z lematu 3. $T_{i,k} = j_1$. Ponieważ rozważamy indeksy w malejącej kolejności, poprawimy wartość $THRESH[k]$ na kolejne $j_p, j_{p-1}, \dots$ aż skończymy na $j_1$.
+
+Ponieważ lemat 1 mówi nam, że wartości w $THRESH$ są w kolejności rosnącej, możemy wyszukiwać $k$ używając wyszukiwania binarnego w czasie $O(log\ n)$. Jeśli oznaczymy ilość wszystkich dopasowań pomiędzy $A$ i $B$, to wykonanie głównej pętli zajmuje $O(r\ log\ n)$. Wyznaczenie list $MATCHLIST$ możemy wykonać sortując $A$ i $B$ pamiętając indeksy i wykonać merge na tych ciągach w łącznym czasie $O(n\ log\ n)$. Potrzebujemy $O(n)$ pamięci na wykorzystywane listy i tablice oraz $O(r)$ pamięci na linki. 

```

Comment:
"Wyznaczenie list $MATCHLIST$ możemy wykonać sortując $A$ i $B$ pamiętając indeksy i wykonać merge na tych ciągach w łącznym czasie $O(n\ log\ n)$."
Could you explain it in a few more words? I am not sure if I get this one exactly right.